include/flang/lower/OpBuilder.h

```diff
@@ -115,24 +116,74 @@ class CharacterOpsBuilder : public OpBuilderWrapper {
   CharacterOpsBuilder(OpBuilderWrapper &b) : OpBuilderWrapper{b} {}
   /// Interchange format to avoid inserting unbox/embox everywhere while
   /// evaluating character expressions.
-  struct CharValue {
-    fir::ReferenceType getReferenceType();
-    fir::CharacterType getCharacterType();
 
-    mlir::Value reference;
+  /// A readable character entity
+  struct CharValue {
+    /// Get fir.char<kind> type with the same kind as value.
+    fir::CharacterType getCharacterType() const;
+    /// Get fir.ref<fir.char<kind>> type.
+    fir::ReferenceType getReferenceType() const;
+    /// Get fir.sequence<? x fir.char<kind>> type.
+    fir::SequenceType getSequenceType() const;
+    /// Get fir.ref<fir.sequence<? x fir.char<kind>>> type.
+    fir::ReferenceType getSequenceRefType() const;
+
+    /// value is a fir.ref<...>
+    bool isRef() const;
+    /// value is not a fir.ref<...>
+    bool isValue() const { return !isRef(); };
+
+    /// Value must be of type:
+    /// - fir.char<kind>
+    /// - fir.sequence<? x fir.char<kind>> type.
+    /// - fir.ref<fir.char<kind>>
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+    mlir::Value value;
     mlir::Value len;
   };
 
+  /// A readable and writable character entity.
+  struct CharRef : CharValue {
+    /// value must be of type:
+    /// - fir.ref<fir.char<kind>>
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+  };
```

Comment:
Given the above, does this `struct` signal that we'll be separating values and references?

Reply:
That was when I thought we would handle `fir.array< fir.char<<>>`, but I will throw that, we will always have have character reference at (box/boxchar) at the interface but maybe for lenght 1 stuff like `ACHAR(20)`.
However, I think we might want to provide an easy way to know if reference are "read-only" or not (that was the purpose of this struct). Something like an attribute on `fir.box_char`/`fir.box` ?

Reply:
Agree that it might be worth considering an annotation for exposing read-only.

Note that there are also global constants, which are implicitly read-only and memory references.